### ISSUE TITLE
Compute CesiumSunSky time zone from longitude

### ISSUE THREAD

**Issue by kring** [CesiumGS/cesium-unreal#1258]
The CesiumSunSky automatically gets its longitude and latitude from the CesiumGeoreference origin, ensuring that the sun position is accurate in the new Unreal coordinate system. However, we currently don't set the time zone. So a common user experience, for example in the [Adding Datasets](https://cesium.com/learn/unreal/unreal-datasets/) tutorial, looks like this:

1. CesiumGeoreference is somewhere, such as New York. The Time Zone property on the CesiumSunSky is set to -5, which is appropriate for New York. The Solar Time is set to 13, so it's about 1pm in New York, and the sun position reflects that.
2. The user wants to build their level in Melbourne. So maybe they add the Melbourne Photogrammetry tileset and double-click it. The tileset is visible, but upside-down because the CesiumGeoreference origin is still back in New York.
3. The user clicks the "Place Georeference Origin Here" button to move the origin near Melbourne.
4. Suddenly the scene goes completely dark. Time Zone -5 plus Solar Time 13 makes about 8AM UTC, which is about 7pm in Melbourne. It's dark there, and the CesiumSunSky accurately reflects that.

The weird thing here is that the user is required to manual enter the time zone. We inherited that from Unreal's SunSky, but there's no good reason for it when we know the longitude. Time zone isn't purely a function of longitude, of course (it's crazy complicated), but when paired with a Solar Time, longitude is really all we need. Solar Time and local time aren't the same thing anyway.

So, I think the user experience could be greatly improved by making CesiumSunSky automatically compute the time zone as a function of the longitude of the georeference origin. That way, after setting the new origin to Melbourne in the scenario above, it would be 1PM in Melbourne (roughly) rather than 1PM in New York.

**Comment by csciguy8** [comment]
I like this idea a lot. It could even mimic time settings we'd see on our computer or phone. A "set automatically" checkbox for time zone, or even solar time.

I encountered a similar user experience problem when setting up a scene in Tokyo. Everything was dark until I set the time zone correctly.

**Comment by dagillespie** [comment]
Timezone from lat/log would be really helpful for a tileset. I found this thread trying to solve the exact problem you're describing.

**Comment by azrogers** [comment]
Probably the main question to answer here is how we can get the current time zone boundary info to figure out which time zone the user is specifying. We can do a really rough estimate based on the rule that time zones change every 15 degrees of longitude, but this will cause issues with the many, many exceptions - Greenland, China, the International Date Line. We could package the [IANA Time Zone Database](https://www.iana.org/time-zones), though it's unclear to me if the boundaries are represented in the database. Alternatively, there's [TimeZoneDB](https://timezonedb.com/) and [Google's Time Zone API](https://developers.google.com/maps/documentation/timezone/overview), but this would require a web request every time we change coordinates. Not the worst thing in the world, considering we're already making many web requests just to stream tiles in the first place.

**Comment by kring** [comment]
I don't think we need accurate time zones. I tried to address this in the top post:

> Time zone isn't purely a function of longitude, of course (it's crazy complicated), but when paired with a Solar Time, longitude is really all we need. Solar Time and local time aren't the same thing anyway.

So a simple linear mapping of time zone to longitude will work just fine I think. It looks like it can even be fractional?

One thing to watch out for is what happens when the georeference moves at runtime, especially as a result of origin rebasing. If it causes the sun to follow you around, that's probably fine. If it causes the sun to jitter, that's bad.

**Comment by j9liu** [comment]
> One thing to watch out for is what happens when the georeference moves at runtime, especially as a result of origin rebasing. If it causes the sun to follow you around, that's probably fine. If it causes the sun to jitter, that's bad.

I might be misreading the intent here, but this will be a toggleable setting (syncing timezone at runtime), right? It makes sense for the Editor experience when using the **Place Origin Here** button, but people may not necessarily want it at runtime. (But perhaps they could enable it if desired)

**Comment by kring** [comment]
> I might be misreading the intent here, but this will be a toggleable setting (syncing timezone at runtime), right? It makes sense for the Editor experience when using the Place Origin Here button, but people may not necessarily want it at runtime.

Yeah, that's a good point, and I agree. It should be Editor-only at least to start, and maybe forever.

**Comment by r-veenstra** [comment]
Being able to call a function / blueprint node to update the time zone at runtime would be useful.  I recently built a project where the user jumped between cesium sub levels on opposite sides of the earth, and I had manually set the time zone when the user flew to the new location to ensure lighting was still appropriate.
